src/grappler/ngraph_optimizer.cc

```diff
@@ -65,18 +66,79 @@ Status NgraphOptimizer::Optimize(tensorflow::grappler::Cluster* cluster,
   // passes become a no-op.
   if (config::IsEnabled() == false ||
       std::getenv("NGRAPH_TF_DISABLE") != nullptr) {
-    NGRAPH_VLOG(0) << "Ngraph is disabled ";
+    NGRAPH_VLOG(0) << "[NGTF-OPTIMIZER] Ngraph is disabled ";
+    graph.ToGraphDef(output);
     return Status::OK();
   }
 
   // Get the nodes to be skipped
-  std::vector<string> fetch_nodes;
+  std::set<string> fetch_nodes;
   for (const string& f : item.fetch) {
-    NGRAPH_VLOG(5) << "Skip fetch node: " << f;
     int pos = f.find(":");
-    fetch_nodes.push_back(f.substr(0, pos));
+    fetch_nodes.insert(f.substr(0, pos));
+  }
+  std::set<string>& skip_these_nodes = fetch_nodes;
+
+  // Rewrite graph to add identity node so the skip node can be encapsulated
+  // as well
+  Graph* input_graph = &graph;
+  for (auto node : input_graph->op_nodes()) {
+    bool fetch_node = false;
+    bool ref_type = false;
+    fetch_node = fetch_nodes.find(node->name()) != fetch_nodes.end();
+    if (fetch_node) {
+      NGRAPH_VLOG(5) << "[NGTF-OPTIMIZER] Fetch Node " << node->name();
+      // Check the number of outputs of the 'fetch_node'
+      // Only move further to create an IdentityN node
+      // if it is greater than 0
+      // Also, make sure that none of the output types is
+      // a ref type
+      if (node->num_outputs()) {
+        std::vector<NodeBuilder::NodeOut> inputs;
+        std::vector<DataType> input_types;
+        for (int i = 0; i < node->num_outputs(); i++) {
+          if (!IsRefType(node->output_type(i))) {
```

Comment:
Can you please add some comments why ref_type is getting special treatment. In case of ref_type would grappler path fail then? Also should we mark this node as skipped, so that its not clustered

Reply:
Maybe in a separate PR we have to consider adding unit tests

Reply:
I did add comments on line 94. I'll add more info.

Reply:
Will be done in the next PR